Add new config_file_service_registration token (#15828)
hashicorp/consul: website/content/docs/agent/config/config-files.mdx

    - `agent_master` ((#acl_tokens_agent_master)) **Renamed in Consul 1.11 to
      [`acl.tokens.agent_recovery`](#acl_tokens_agent_recovery).**

    - `config_file_service_registration` ((#acl_tokens_config_file_service_registration)) - The ACL token this agent
      uses to register services and checks from [service definitions](/docs/discovery/services) and

>       [check definitions](/docs/discovery/checks) foudn in configuration files or in configuration
>       strings passed to the agent using the `-hcl` flag.

Reviewer: Does registration also happen on reload?

If so, we could do something like:
```suggestion
      [check definitions](/docs/discovery/checks) loaded by the agent on startup and reload.
```

>       strings passed to the agent using the `-hcl` flag.

Reviewer: nit: If we keep the `-hcl` flag description, would we also need to include `-json`?

>       strings passed to the agent using the `-hcl` flag.

Author: From what I can tell, `-json` doesn't exist. The `consul agent` command only has `-hcl` for config fragments.

>       strings passed to the agent using the `-hcl` flag.

Reviewer 2: The `-hcl` flag enables operators to specify agent configuration values on the CLI. There is currently no equivalent `-json` flag for allowing agent configuration to be provided in JSON format. If we wanted to support that, it would be a new feature that requires additional development.

>       strings passed to the agent using the `-hcl` flag.

Reviewer: I don't want to suggest we support that. For some reason I thought I had seen an invocation of a Consul agent with that recently, but it seems like I misremembered (e.g., perhaps it was a config file being passed in with JSON format).

>       strings passed to the agent using the `-hcl` flag.

Author: I think I prefer to be more elaborate / specific here to help reduce confusion?

If we say "configuration passed at startup", I feel like "startup" leaves room for interpretation. Is sending an HTTP request passing configuration to the agent? Does that include if I send a service registration request while the agent is "starting up"? 

I wanted to be clear about what services/checks the token is specifically used for (those services/checks in files or `-hcl` config fragments).

      If an inline token is defined in the service or check definition, then the inline token is
      used to register that service or check instead. If the `config_file_service_registration` token is not
      defined and if a service or check has no inline token, then the agent uses the
      [`default`](#acl_tokens_default) token to register the service or check.

      This token needs permissions for all services and checks this agent needs to register. For
      example, if there are multiple service definitions in configuration files, then the
      `config_file_service_registration` token needs multiple `service:write` permissions in order for

>       the agent to register those services.

Reviewer 3: On that note, what happens if a `config_file_service_registration` token has permissions for a partial set of services? Does it fail to write all services or does it skip only the service with the missing perm?

Could make the behavior clear in the docs here.

>       the agent to register those services.

Author: I rewrote this using two named services "A" and "B" as an example. I also elaborated a bit more on the failure case (maybe too much?). Let me know what you think!
